rp2040-hal/src/timer.rs

```diff
@@ -84,27 +94,85 @@ impl Timer {
     }
     /// Retrieve a reference to alarm 0. Will only return a value the first time this is called
     pub fn alarm_0(&mut self) -> Option<Alarm0> {
-        take_alarm(1 << 0).then_some(Alarm0(PhantomData))
+        take_alarm(1 << 0).then_some(Alarm0(*self))
     }
 
     /// Retrieve a reference to alarm 1. Will only return a value the first time this is called
     pub fn alarm_1(&mut self) -> Option<Alarm1> {
-        take_alarm(1 << 1).then_some(Alarm1(PhantomData))
+        take_alarm(1 << 1).then_some(Alarm1(*self))
     }
 
     /// Retrieve a reference to alarm 2. Will only return a value the first time this is called
     pub fn alarm_2(&mut self) -> Option<Alarm2> {
-        take_alarm(1 << 2).then_some(Alarm2(PhantomData))
+        take_alarm(1 << 2).then_some(Alarm2(*self))
     }
 
     /// Retrieve a reference to alarm 3. Will only return a value the first time this is called
     pub fn alarm_3(&mut self) -> Option<Alarm3> {
-        take_alarm(1 << 3).then_some(Alarm3(PhantomData))
+        take_alarm(1 << 3).then_some(Alarm3(*self))
+    }
+
+    /// Pauses execution for at minimum `us` microseconds.
+    fn delay_us_internal(&self, mut us: u32) {
+        let mut start = self.get_counter_low();
+        // If we knew that the loop ran at least once per timer tick,
+        // this could be simplified to:
+        // ```
+        // while timer.timelr.read().bits().wrapping_sub(start) <= us {
+        //     cortex_m::asm::nop();
+        // }
+        // ```
+        // However, due to interrupts, for `us == u32::MAX`, we could
+        // miss the moment where the loop should terminate if the loop skips
+        // a timer tick.
+        loop {
+            let now = self.get_counter_low();
+            let waited = now.wrapping_sub(start);
+            if waited >= us {
+                break;
+            }
+            start = now;
+            us -= waited;
+        }
     }
 }
 
-// safety: all write operations are synchronised and all reads are atomic
-unsafe impl Sync for Timer {}
+macro_rules! impl_delay_traits {
+    ($($t:ty),+) => {
+        $(
+        impl embedded_hal::blocking::delay::DelayUs<$t> for Timer {
+            fn delay_us(&mut self, mut us: $t) {
+                #![allow(unused_comparisons)]
+                assert!(us >= 0); // Only meaningful for i32
+                while <$t>::MAX as u64 > u32::MAX as u64 && us > u32::MAX as $t {
+                    self.delay_us_internal(u32::MAX);
+                    us -= u32::MAX as $t;
+                }
```

Comment:
I really find this convoluted for a loop that's only present when we're implementing support for u64.
I think it would be more readable and less error-prone to have this in its own impl block.

Reply:
Fair. It's one thing that the optimizer removes the loop, but that doesn't help readability. I'll implement the u64 version separately.

Reply:
Or should we skip the u64 implementation entirely? A blocking delay of more than an hour is usually just wrong. And if someone really needs it for some reason, it's not difficult to implement it locally:
```
 use embedded_hal::blocking::delay::DelayUs;                                                                                                                                                                     
                                                                                                                                                                                                                 
 pub struct DelayU64<T>(T);                                                                                                                                                                                      
                                                                                                                                                                                                                 
 impl<T> DelayU64<T> {                                                                                                                                                                                           
     pub fn new(delay: T) -> Self {                                                                                                                                                                              
         Self(delay)                                                                                                                                                                                             
     }                                                                                                                                                                                                           
 }                                                                                                                                                 
                                                                                                                                                                                                                 
 impl<T> DelayUs<u64> for DelayU64<T>                                                                                                                                                                            
 where                                                                                                                                                                                                           
     T: DelayUs<u32>,                                                                                                                                                                                            
 {                                                                                                                                                                                                               
     fn delay_us(&mut self, mut us: u64) {                                                                                                                                                                 
         while us > u32::MAX as u64 {                                                                                                                                                                
             self.0.delay_us(u32::MAX);                                                                                                                                                                          
             us -= u32::MAX as u64;                                                                                             
         }                                                                                                                      
         self.0.delay_us(us as u32);                                                                                            
     }                                                                                                                          
 } 
 ```
 
 And in case anyone really misses these functions, we can always add them later.